core/jvm/src/test/scala/matryoshka/instances/fixedpoint/partial.scala

```diff
@@ -16,37 +16,95 @@
-  implicit def eitherArbitrary[A: Arbitrary]:
-      Arbitrary ~> λ[β => Arbitrary[A \/ β]] =
-    new (Arbitrary ~> λ[β => Arbitrary[A \/ β]]) {
-      def apply[β](arb: Arbitrary [β]) =
-        Arbitrary(Gen.oneOf(
-          Arbitrary.arbitrary[A].map(-\/(_)),
-          arb.arbitrary.map(\/-(_))))
-    }
+// TODO: Figure out how to implement Monads generically for fixed-point types,
+//       and get them scoped automatically (companion for a type alias doesn’t
+//       work)
+import Partial.monad
 
+class PartialSpec extends Specification with ScalazMatchers with ScalaCheck with CheckAll {
   /** For testing cases that should work with truly diverging functions. */
   def sometimesNeverGen[A: Arbitrary]: Gen[Partial[A]] =
     Gen.oneOf(Arbitrary.arbitrary[Partial[A]], Gen.const(Partial.never[A]))
 
   "Partial" should {
     "satisfy relevant laws" in {
-      checkAll(equal.laws[Partial[Int]](Partial.equal, implicitly))
-      checkAll(monad.laws[Partial](Partial.monad, implicitly, implicitly, implicitly, Partial.equal))
+      checkAll(Props.equal.laws[Partial[Int]](Partial.equal, implicitly))
+      checkAll(Props.monad.laws[Partial](implicitly, implicitly, implicitly, implicitly, Partial.equal))
+    }
+  }
+
+  // https://en.wikipedia.org/wiki/McCarthy_91_function
+  def mc91(n: Int): Partial[Int] =
+    if (n > 100) Partial.now(n - 10)
+    else mc91(n + 11) >>= mc91
+
+  "never" should {
+    "always have more steps" in {
+      Partial.never[Int].runFor(1000) must beRightDisjunction
+    }
+  }
+
+  "runFor" should {
+    "return now immediately" in {
+      Partial.now(13).runFor(0) must beLeftDisjunction(13)
+    }
+
+    "return a value when it runs past the end" in {
+      Partial.later(Partial.now(7)).runFor(300000) must beLeftDisjunction(7)
+    }
+
+    "return after multiple runs" ! prop { (a: Conat, b: Conat) =>
+      val (ai, bi) = (a.cata(height), b.cata(height))
+      bi > 0 ==> {
+        val first = (a + b).transAna(Partial.delay(27)).runFor(ai)
+        first must beRightDisjunction
+        first.flatMap(_.runFor(bi)) must beLeftDisjunction(27)
+      }
+    }
+
+    "still pending one short" ! prop { (a: Conat) =>
+      val ai = a.cata(height)
+      ai > 0 ==> {
+        val first = a.transAna(Partial.delay(27)).runFor(ai - 1)
+        first must beRightDisjunction
+        first.flatMap(_.runFor(1)) must beLeftDisjunction(27)
+      }
+    }
+
+    "return exactly at the end" ! prop { (i: Conat) =>
+      i.transAna(Partial.delay(4)).runFor(i.cata(height)) must
+        beLeftDisjunction(4)
+    }
+  }
+
+  "unsafePerformSync" should {
+    "return now immediately" in {
+      Partial.now(12).unsafePerformSync must equal(12)
+    }
+
+    "return a value when it gets to the end" in {
+      Partial.later(Partial.later(Partial.now(3))).unsafePerformSync must
+        equal(3)
+    }
+
+    // TODO: Should work with any Int, but stack overflows on big negatives.
```

Comment:
Yeah, not sure how to make `mc91` tailrec, maybe `Trampoline` to avoid overflow? (No idea if that is possible with `Partial` either!)